nucypher-core/src/access_control.rs

```diff
@@ -0,0 +1,209 @@
+use alloc::boxed::Box;
+use alloc::string::String;
+use alloc::vec::Vec;
+
+use ferveo::api::{encrypt, Ciphertext, DkgPublicKey, SecretBox};
+use ferveo::Error;
+use serde::{Deserialize, Serialize};
+use umbral_pre::serde_bytes;
+
+use crate::conditions::Conditions;
+use crate::versioning::{
+    messagepack_deserialize, messagepack_serialize, ProtocolObject, ProtocolObjectInner,
+};
+
+/// Authenticated data for encrypted data.
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct AuthenticatedData {
+    /// The public key for the encrypted data
+    pub public_key: DkgPublicKey,
+
+    /// The conditions associated with the encrypted data
+    pub conditions: Option<Conditions>,
+}
+
+impl AuthenticatedData {
+    /// Creates a new access control policy.
+    pub fn new(public_key: &DkgPublicKey, conditions: Option<&Conditions>) -> Self {
+        AuthenticatedData {
+            public_key: *public_key,
+            conditions: conditions.cloned(),
+        }
+    }
+
+    /// Return the aad.
+    pub fn aad(&self) -> Box<[u8]> {
+        let public_key_bytes = self.public_key.to_bytes().unwrap();
+        let condition_bytes = self.conditions.as_ref().unwrap().as_ref().as_bytes();
+        let mut result = Vec::with_capacity(public_key_bytes.len() + condition_bytes.len());
+        result.extend(public_key_bytes);
+        result.extend(condition_bytes);
+        result.into_boxed_slice()
+    }
+}
+
+impl PartialEq for AuthenticatedData {
+    fn eq(&self, other: &Self) -> bool {
+        self.public_key.to_bytes().unwrap() == other.public_key.to_bytes().unwrap()
+            && self.conditions == other.conditions
+    }
+}
+
+impl Eq for AuthenticatedData {}
```

Comment:
Can we `#[derive(Eq)]` on `AuthenticatedData` instead? I assume there is a reason deriving `Eq` in this particular way.

Reply:
I may have mistakenly thought that you have to do this if you implemented `PartialEq`. Was that mistaken? In any case, it's not a blocker right...?

Reply:
While working on #77 I realized that automatic derivation of `Eq` is not possible since `DkgPublicKey` from `ferveo` doesn't implement it either. 

So it makes sense to provide an implementation stub `impl Eq {}` as a workaround. We need `Eq` do implement `ProtocolObjectInner` for `AccessControlPolicy`. Tracking this here: https://github.com/nucypher/ferveo/issues/157